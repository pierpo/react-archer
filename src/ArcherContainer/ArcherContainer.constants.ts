export const shapeDefaultProp = {
  arrow: {
    arrowLength: 10,
    arrowThickness: 6,
  },
  circle: {
    radius: 2,
    fillColor: '#f00',
    strokeColor: '#0ff',
    strokeWidth: 1,
  },
};
