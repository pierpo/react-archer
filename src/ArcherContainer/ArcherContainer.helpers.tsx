import {
  EntityRelationType,
  LineType,
  ShapeType,
  SourceToTargetType,
  ValidShapeTypes,
} from '../types';
import { SourceToTargetsArrayType } from './ArcherContainer.types';

const possibleShapes: Array<ValidShapeTypes> = ['arrow', 'circle'];

export const getEndShapeFromStyle = (shapeObj: LineType, shape: ShapeType) => {
  if (shapeObj.endShape) {
    const validShape = (Object.keys(shapeObj.endShape) as ValidShapeTypes[]).find((key) => {
      return possibleShapes.includes(key);
    });

    if (validShape) {
      return validShape;
    }
  }

  if (shape.arrow) {
    return 'arrow';
  } else if (shape.circle) {
    return 'circle';
  }

  return possibleShapes[0];
};

export const getSourceToTargets = (
  sourceToTargetsMap: Record<string, SourceToTargetsArrayType>,
): SourceToTargetType[] => {
  // Object.values is unavailable in IE11
  const jaggedSourceToTargets = Object.keys(sourceToTargetsMap).map(
    (key: string) => sourceToTargetsMap[key],
  );
  // Flatten
  return ([] as SourceToTargetType[]).concat
    .apply([], jaggedSourceToTargets)
    .sort((a, b) => a.order - b.order);
};

export const createShapeObj = (style: LineType, endShape: ShapeType) => {
  const chosenEndShape = getEndShapeFromStyle(style, endShape);
  const shapeObjMap = {
    arrow: () => ({
      arrow: {
        ...endShape?.arrow,
        ...style.endShape?.arrow,
      },
    }),
    circle: () => ({
      circle: {
        ...endShape?.circle,
        ...style.endShape?.circle,
      },
    }),
  };
  return shapeObjMap[chosenEndShape]();
};

/** Generates an id for an arrow marker
 * Useful to have one marker per arrow so that each arrow
 * can have a different color!
 * */
export const getMarkerId = (
  uniqueId: string,
  source: EntityRelationType,
  target: EntityRelationType,
): string => {
  return `${uniqueId}${source.id}${target.id}`;
};
