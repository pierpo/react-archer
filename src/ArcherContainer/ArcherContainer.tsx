import React, { useCallback, useImperativeHandle, useMemo, useRef, useState } from 'react';
import ResizeObserver from 'resize-observer-polyfill';
import { SourceToTargetType } from '../types';
import { ArcherContainerContext, ArcherContainerContextProvider } from './ArcherContainer.context';
import {
  SourceToTargetsArrayType,
  ArcherContainerProps,
  ArcherContainerHandle,
} from './ArcherContainer.types';
import { SvgArrows } from './components/SvgArrows';
import { endShapeDefaultProp } from './ArcherContainer.constants';
import { ArrowMarkers } from './components/Markers';
import { useObserveElements, useResizeListener } from './ArcherContainer.hooks';

const defaultSvgContainerStyle = {
  position: 'absolute',
  width: '100%',
  height: '100%',
  top: 0,
  left: 0,
  pointerEvents: 'none',
};

const ArcherContainer = React.forwardRef<ArcherContainerHandle, ArcherContainerProps>(
  (
    {
      endShape = endShapeDefaultProp,
      strokeColor = '#f00',
      strokeWidth = 2,
      svgContainerStyle = {},
      noCurves,
      children,
      className,
      endMarker,
      lineStyle,
      offset,
      startMarker,
      strokeDasharray,
      ignoreNotFoundWarnings,
      style,
    }: ArcherContainerProps,
    archerContainerRef,
  ) => {
    const [refs, setRefs] = useState<Record<string, HTMLElement>>({});
    const [sourceToTargetsMap, setSourceToTargetsMap] = useState<
      Record<string, SourceToTargetsArrayType>
    >({});
    const observer = useRef<ResizeObserver>(
      new ResizeObserver(() => {
        refreshScreen();
      }),
    ).current;

    const parent = useRef<HTMLDivElement>(null);

    const [, updateState] = React.useState<{}>();

    const uniqueId = useRef<string>(`arrow${Math.random().toString().slice(2)}`).current;

    useImperativeHandle(
      archerContainerRef,
      (): ArcherContainerHandle => ({
        refreshScreen,
        arrowMarkerUniquePrefix: uniqueId,
      }),
    );

    /**
     * Use this to recompute all the arrow positions. Useful if arrows do not properly rerender
     * after the viewport or some elements moved.
     */
    const refreshScreen = React.useCallback(() => updateState({}), []);

    const _registerTransitions = useCallback(
      (elementId: string, newSourceToTargets: SourceToTargetType[]): void => {
        setSourceToTargetsMap((previousValue) => ({
          ...previousValue,
          [elementId]: newSourceToTargets,
        }));
      },
      [],
    );

    const _unregisterTransitions = useCallback((elementId: string): void => {
      setSourceToTargetsMap((previousValue) => {
        const sourceToTargetsMapCopy = { ...previousValue };
        delete sourceToTargetsMapCopy[elementId];
        return sourceToTargetsMapCopy;
      });
    }, []);

    const _registerChild = useCallback((id: string, ref: HTMLElement): void => {
      setRefs((currentRefs) => {
        if (currentRefs[id] === ref) {
          return currentRefs;
        }

        return {
          ...currentRefs,
          [id]: ref,
        };
      });
    }, []);

    const _unregisterChild = useCallback((id: string): void => {
      setRefs((currentRefs) => {
        const newRefs = { ...currentRefs };
        delete newRefs[id];
        return newRefs;
      });
    }, []);

    const _svgContainerStyle = useMemo(
      (): Record<string, any> => ({
        ...defaultSvgContainerStyle,
        ...svgContainerStyle,
      }),
      [svgContainerStyle],
    );

    let newChildren: React.ReactNode | null | undefined;

    if (typeof children === 'function') {
      newChildren = children(ArcherContainerContext);
    } else {
      newChildren = children;
    }

    useResizeListener(refreshScreen);

    useObserveElements(refs, observer);

    const contextValue = useMemo(
      () => ({
        registerTransitions: _registerTransitions,
        unregisterTransitions: _unregisterTransitions,
        registerChild: _registerChild,
        unregisterChild: _unregisterChild,
      }),
      [_registerTransitions, _unregisterTransitions, _registerChild, _unregisterChild],
    );

    return (
      <ArcherContainerContextProvider value={contextValue}>
        <div style={{ ...style, position: 'relative' }} className={className}>
          <svg style={_svgContainerStyle}>
            <defs>
              <ArrowMarkers
                endShape={endShape}
                sourceToTargetsMap={sourceToTargetsMap}
                strokeColor={strokeColor}
                uniqueId={uniqueId}
              />
            </defs>
            <SvgArrows
              startMarker={startMarker}
              endMarker={endMarker}
              endShape={endShape}
              strokeColor={strokeColor}
              strokeWidth={strokeWidth}
              strokeDasharray={strokeDasharray}
              noCurves={noCurves}
              lineStyle={lineStyle}
              offset={offset}
              ignoreNotFoundWarnings={ignoreNotFoundWarnings}
              parentCurrent={parent.current}
              refs={refs}
              uniqueId={uniqueId}
              sourceToTargetsMap={sourceToTargetsMap}
            />
          </svg>

          <div
            style={{
              height: '100%',
            }}
            ref={parent}
          >
            {newChildren}
          </div>
        </div>
      </ArcherContainerContextProvider>
    );
  },
);

export default ArcherContainer;
