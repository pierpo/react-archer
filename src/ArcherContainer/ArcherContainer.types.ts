import { ShapeType, ValidLineStyles, SourceToTargetType } from '../types';
import { ArcherContainerContextType } from './ArcherContainer.context';

type FunctionChild = (context: React.Context<ArcherContainerContextType | null>) => React.ReactNode;

export type ArcherContainerProps = {
  /**
   * A color string
   *
   * @example '#ff0000'
   */
  strokeColor?: string;

  /**
   * A size in px
   */
  strokeWidth?: number;

  /**
   * A string representing an array of sizes
   * See https://www.w3schools.com/graphics/svg_stroking.asp
   */
  strokeDasharray?: string;

  style?: React.CSSProperties;

  /**
   * Style of the SVG container element. Useful if you want to add a z-index to your SVG container to draw the arrows under your elements, for example.
   */
  svgContainerStyle?: React.CSSProperties;

  className?: string;

  /**
   * Optional number for space between element and start/end of stroke
   */
  offset?: number;

  /**
   * Customize the end shape of the line. Defaults to a traditional "arrow" (triangle) shape.
   */
  endShape?: ShapeType;

  /**
   * Set this to true of you want to render a marker at the start of the line
   */
  startMarker?: boolean;

  /**
   * Set this to false of you do not want to render a marker at the end of the line
   */
  endMarker?: boolean;

  /**
   * Define how the line is drawn, grid for angles, straight for direct line and curve for curves
   */

  lineStyle?: ValidLineStyles;

  /**
   * Set this to true if you want angles instead of curves
   */
  noCurves?: boolean;

  /**
   * Set this to true if you want to ignore warnings when a target or source is not found
   */
  ignoreNotFoundWarnings?: boolean;

  children?: React.ReactNode | FunctionChild;
};

export type ArcherContainerRef = {
  /**
   * Use this to recompute all the arrow positions. Useful if arrows do not properly rerender
   * after the viewport or some elements moved.
   */
  refreshScreen: () => void;
  /** Only for internal testing purposes */
  arrowMarkerUniquePrefix: string;
};

// Keep this name for retrocompatibility
// should be a breaking change of 4.0.0
export type ArcherContainerHandle = ArcherContainerRef;

export type SourceToTargetsArrayType = SourceToTargetType[];
