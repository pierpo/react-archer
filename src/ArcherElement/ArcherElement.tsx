import React, { useCallback, useContext, useLayoutEffect, useMemo, useRef } from 'react';
import { ArcherContainerContext } from '../ArcherContainer/ArcherContainer.context';
import { RelationType } from '../types';
import { encodeId } from '../utils/encodeId';
import { useDeepCompareEffect } from '../utils/useDeepCompareEffect';
import { generateSourceToTarget, assertContextExists } from './ArcherElement.helpers';

type ArcherElementProps = {
  /**
   * The id that will identify the Archer Element.
   */
  id: string;
  relations?: Array<RelationType>;
  children: React.ReactElement;
};

const ArcherElement = ({ id, relations = [], children }: ArcherElementProps) => {
  const encodedId = useMemo(() => encodeId(id), [id]);
  const context = useContext(ArcherContainerContext);
  const archerRef = useRef<HTMLElement>();

  const registerTransitions = useCallback(
    (newRelations: Array<RelationType>) => {
      const newSourceToTarget = generateSourceToTarget(encodedId, newRelations);

      assertContextExists(context);

      context.registerTransitions(encodedId, newSourceToTarget);
    },
    [context, encodedId],
  );

  const unregisterTransitions = useCallback(() => {
    assertContextExists(context);

    context.unregisterTransitions(encodedId);
  }, [context, encodedId]);

  const registerChild = useCallback(
    (newRef: HTMLElement | undefined) => {
      if (!newRef) return;

      assertContextExists(context);

      context.registerChild(encodedId, newRef);
    },
    [encodedId, context],
  );

  const unregisterChild = useCallback(() => {
    assertContextExists(context);

    context.unregisterChild(encodedId);
  }, [context, encodedId]);

  useLayoutEffect(() => {
    registerChild(archerRef.current);

    return () => unregisterChild();
  }, [registerChild, unregisterChild]);

  useDeepCompareEffect(() => {
    registerTransitions(relations);

    return () => unregisterTransitions();
  }, [registerTransitions, relations, unregisterTransitions]);

  // Check that we only have one child to ArcherElement
  React.Children.only(children);
  // Now, we'll render this child by getting its ref. The ref will be used to compute the element's position.
  // I'm pretty sure there's a cleaner way to get the ref of the child... feel free to suggest it!
  const child = children;
  return React.cloneElement(child, {
    ...child.props,
    ref: (node: HTMLElement) => {
      archerRef.current = node;

      // @ts-expect-error could not find a proper type for Children
      const ref = child.ref as React.MutableRefObject<any> | React.LegacyRef<any>;

      if (!ref) {
        return;
      }

      if (typeof ref === 'string') {
        console.error('[React Archer] Legacy string refs are not supported');
        return;
      }

      if (typeof ref === 'function') {
        ref(node);
        return;
      }

      if (ref && ref.current !== undefined) {
        // @ts-expect-error LegacyRef cast above is annoying - it marks the ref as non mutable
        ref.current = node;
      }
    },
  });
};

export default ArcherElement;
