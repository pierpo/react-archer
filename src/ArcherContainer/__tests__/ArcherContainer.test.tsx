import React from 'react';
import { render } from '@testing-library/react';
import ArcherContainer from '../ArcherContainer';
import ArcherElement from '../../ArcherElement/ArcherElement';
import { act } from 'react-dom/test-utils';

const originalConsoleWarn = console.warn;

describe('ArcherContainer', () => {
  beforeEach(() => {
    console.warn = originalConsoleWarn;
  });

  it('should render given children', async () => {
    const screen = render(
      <ArcherContainer>
        <div>child</div>
      </ArcherContainer>,
    );
    await screen.findByText('child');
  });

  describe('rendering an svg with the marker element used to draw an svg arrow', () => {
    it('should render simple elements', async () => {
      const screen = render(
        <ArcherContainer>
          <ArcherElement
            id="elem-left"
            relations={[{ sourceAnchor: 'left', targetAnchor: 'right', targetId: 'elem-right' }]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="elem-right">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );
      expect(screen.baseElement).toMatchSnapshot();
    });

    it('should render the arrow with an arrow end by default', () => {
      const screen = render(
        <ArcherContainer>
          <ArcherElement id="elem-left">
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement
            id="elem-right"
            relations={[{ sourceAnchor: 'right', targetAnchor: 'left', targetId: 'elem-left' }]}
          >
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );

      expect(screen.baseElement).toMatchSnapshot();
    });

    it('should render the arrow with a circle end when provided', () => {
      const screen = render(
        <ArcherContainer>
          <ArcherElement
            id="elem-left"
            relations={[
              {
                sourceAnchor: 'left',
                targetAnchor: 'right',
                targetId: 'elem-right',
                style: {
                  endShape: {
                    circle: {
                      radius: 11,
                      strokeWidth: 2,
                      strokeColor: 'tomato',
                      fillColor: '#c0ffee',
                    },
                  },
                },
              },
            ]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="elem-right">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );
      expect(screen.baseElement).toMatchSnapshot();
    });

    it('should render the arrow on both ends', () => {
      const screen = render(
        <ArcherContainer startMarker>
          <ArcherElement
            id="elem-left"
            relations={[
              {
                sourceAnchor: 'left',
                targetAnchor: 'right',
                targetId: 'elem-right',
              },
            ]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="elem-right">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );
      expect(screen.baseElement).toMatchSnapshot();
    });

    it('should render an arrow with className', () => {
      const screen = render(
        <ArcherContainer startMarker>
          <ArcherElement
            id="elem-left"
            relations={[
              {
                className: 'blink',
                sourceAnchor: 'left',
                targetAnchor: 'right',
                targetId: 'elem-right',
              },
            ]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="elem-right">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );
      expect(screen.baseElement).toMatchSnapshot();
    });

    it('should render the arrows with labels', () => {
      const screen = render(
        <ArcherContainer startMarker endMarker={false}>
          <ArcherElement
            id="elem-left"
            relations={[
              {
                sourceAnchor: 'left',
                targetAnchor: 'right',
                targetId: 'elem-right',
                label: <div>this is a fine label</div>,
              },
            ]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="elem-right">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );
      expect(screen.baseElement).toMatchSnapshot();
    });

    it('should render the arrow only at start', () => {
      const screen = render(
        <ArcherContainer startMarker endMarker={false}>
          <ArcherElement
            id="elem-left"
            relations={[
              {
                sourceAnchor: 'left',
                targetAnchor: 'right',
                targetId: 'elem-right',
              },
            ]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="elem-right">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );
      expect(screen.baseElement).toMatchSnapshot();
    });

    it('should render no arrow if id is not found', () => {
      console.warn = jest.fn();

      const screen = render(
        <ArcherContainer startMarker endMarker={false}>
          <ArcherElement
            id="elem-left"
            relations={[
              {
                sourceAnchor: 'left',
                targetAnchor: 'right',
                targetId: 'oh no',
              },
            ]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="oops">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );

      expect(screen.baseElement).toMatchSnapshot();
      expect(console.warn).toHaveBeenCalledWith(
        expect.stringMatching('Could not find target element'),
      );
    });

    it('should render no arrow if id is not found and not raise a console.warn', () => {
      console.warn = jest.fn();

      const screen = render(
        <ArcherContainer startMarker endMarker={false} ignoreNotFoundWarnings>
          <ArcherElement
            id="elem-left"
            relations={[
              {
                sourceAnchor: 'left',
                targetAnchor: 'right',
                targetId: 'oh no',
              },
            ]}
          >
            <div>element 1</div>
          </ArcherElement>
          <ArcherElement id="oops">
            <div>element 2</div>
          </ArcherElement>
        </ArcherContainer>,
      );

      expect(screen.baseElement).toMatchSnapshot();
      expect(console.warn).not.toHaveBeenCalledWith(
        expect.stringMatching('Could not find target element'),
      );
    });
  });

  describe('Event Listeners', () => {
    it('should move elements on window resize', () => {
      const screen = render(
        <ArcherContainer startMarker endMarker={false}>
          <div
            style={{
              display: 'flex',
              justifyContent: 'space-around',
              alignItems: 'stretch',
            }}
          >
            <ArcherElement
              id="elem-left"
              relations={[
                {
                  sourceAnchor: 'left',
                  targetAnchor: 'right',
                  targetId: 'elem-right',
                },
              ]}
            >
              <div>element 1</div>
            </ArcherElement>
            <ArcherElement id="elem-right">
              <div>element 2</div>
            </ArcherElement>
          </div>
        </ArcherContainer>,
      );
      const element2 = screen.getByText('element 2');

      const pathBefore = document.querySelector('path');
      expect(pathBefore?.attributes.getNamedItem('d')?.value).toEqual('M0,0 L0,6 L10,3 z');

      // Sadly all arrows return the same value for getBoundingClientRect
      // let's change the implementation, trigger a resize, then see that the values did change
      element2.getBoundingClientRect = jest.fn().mockReturnValue({
        width: 20,
        height: 20,
        top: 0,
        left: 0,
        right: 20,
        bottom: 20,
      });
      act(() => window.resizeTo(500, 500));
      const path = document.querySelector('path');
      expect(path?.attributes.getNamedItem('d')?.value).toEqual('M0,0 L0,6 L10,3 z');
    });
  });
});
