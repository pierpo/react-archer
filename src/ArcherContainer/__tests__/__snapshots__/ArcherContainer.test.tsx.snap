// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render an arrow with className 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftelem-right"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
        <g
          class="blink"
        >
          <path
            d="M-20,0 C0,0 0,0 20,0"
            marker-end="url(#arrow00001elem-leftelem-right)"
            marker-start="url(#arrow00001elem-leftelem-right)"
            style="fill: none; stroke: #f00; stroke-width: 2;"
          />
        </g>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render no arrow if id is not found 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftoh_20no"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render no arrow if id is not found and not raise a console.warn 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftoh_20no"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render simple elements 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftelem-right"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
        <g>
          <path
            d="M0,0 C10,0 10,0 20,0"
            marker-end="url(#arrow00001elem-leftelem-right)"
            style="fill: none; stroke: #f00; stroke-width: 2;"
          />
        </g>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render the arrow on both ends 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftelem-right"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
        <g>
          <path
            d="M-20,0 C0,0 0,0 20,0"
            marker-end="url(#arrow00001elem-leftelem-right)"
            marker-start="url(#arrow00001elem-leftelem-right)"
            style="fill: none; stroke: #f00; stroke-width: 2;"
          />
        </g>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render the arrow only at start 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftelem-right"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
        <g>
          <path
            d="M-20,0 C-10,0 -10,0 0,0"
            marker-start="url(#arrow00001elem-leftelem-right)"
            style="fill: none; stroke: #f00; stroke-width: 2;"
          />
        </g>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render the arrow with a circle end when provided 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftelem-right"
            markerHeight="44"
            markerUnits="strokeWidth"
            markerWidth="44"
            orient="auto-start-reverse"
            refX="24"
            refY="22"
          >
            <circle
              cx="22"
              cy="22"
              fill="#c0ffee"
              r="11"
              stroke="tomato"
              stroke-width="2"
            />
          </marker>
        </defs>
        <g>
          <path
            d="M0,0 C11,0 11,0 22,0"
            marker-end="url(#arrow00001elem-leftelem-right)"
            style="fill: none; stroke: #f00; stroke-width: 2;"
          />
        </g>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render the arrow with an arrow end by default 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-rightelem-left"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
        <g>
          <path
            d="M0,0 C-10,0 -10,0 -20,0"
            marker-end="url(#arrow00001elem-rightelem-left)"
            style="fill: none; stroke: #f00; stroke-width: 2;"
          />
        </g>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ArcherContainer rendering an svg with the marker element used to draw an svg arrow should render the arrows with labels 1`] = `
<body>
  <div>
    <div
      style="position: relative;"
    >
      <svg
        style="position: absolute; width: 100%; height: 100%; top: 0px; left: 0px; pointer-events: none;"
      >
        <defs>
          <marker
            id="arrow00001elem-leftelem-right"
            markerHeight="6"
            markerUnits="strokeWidth"
            markerWidth="10"
            orient="auto-start-reverse"
            refX="0"
            refY="3"
          >
            <path
              d="M0,0 L0,6 L10,3 z"
              fill="#f00"
            />
          </marker>
        </defs>
        <g>
          <path
            d="M-20,0 C-10,0 -10,0 0,0"
            marker-start="url(#arrow00001elem-leftelem-right)"
            style="fill: none; stroke: #f00; stroke-width: 2;"
          />
          <foreignobject
            height="1"
            style="overflow: visible; pointer-events: none;"
            width="20"
            x="-20"
            y="0"
          >
            <div
              style="position: absolute; left: 50%; top: 50%; transform: translateX(-50%) translateY(-50%); pointer-events: all;"
            >
              <div>
                <div>
                  this is a fine label
                </div>
              </div>
            </div>
          </foreignobject>
        </g>
      </svg>
      <div
        style="height: 100%;"
      >
        <div>
          element 1
        </div>
        <div>
          element 2
        </div>
      </div>
    </div>
  </div>
</body>
`;
